dbcrossbar/src/clouds/gcloud/client.rs

```diff
@@ -222,6 +222,87 @@ impl Client {
         }
     }
 
+    /// Make an HTTP POST request and return the response.
+    ///
+    /// This may POST the request multiple times, which may cause an action to be
+    /// performed multiple times. The caller is responsible for ensuring that the action
+    /// is idempotent (perhaps by using `IF NOT EXISTS` for an SQL operation), and
+    /// being ready to handle the case where the underlying action succeeds, but the
+    /// server still returns an error.
+    async fn post_with_retry_helper<Body>(
+        &self,
+        url: &Url,
+        body: &Body,
+    ) -> Result<reqwest::Response, ClientError>
+    where
+        Body: fmt::Debug + Serialize + Sync + Send,
+    {
+        trace!("POST {}", url);
+        let token = self.token().await?;
+        let wait_options = WaitOptions::default()
+            .backoff_type(BackoffType::Exponential)
+            .retry_interval(Duration::from_secs(1))
+            // Don't retry too much because we're probably classifying some
+            // permanent errors as temporary.
+            .allowed_errors(5);
+        wait(&wait_options, move || {
+            let token = token.clone();
+            async move {
+                let resp_result = self
+                    .client
+                    .post(url.as_str())
+                    .bearer_auth(token.as_str())
+                    .json(body)
+                    .send()
+                    .await;
+                match resp_result {
+                    // The HTTP request failed outright, because of something
+                    // like a DNS error or whatever.
+                    Err(err) => {
+                        // Request and timeout errors look like the kind of
+                        // things we should probably retry. But this is based on
+                        // guesswork not experience.
+                        let temporary = err.is_request() || err.is_timeout();
+                        let err: Error = err.into();
+                        let err: ClientError =
+                            err.context(format!("could not POST {}", url)).into();
+                        if temporary {
+                            WaitStatus::FailedTemporarily(err)
```

Comment:
We're returning temporary failures on a bunch of errors we've never seen in practice, including request errors which should almost certainly be hard failures.

Reply:
Hmm, no, these aren't 4xx request errors, these are "we couldn't talk to the server using HTTP request errors". Different beasts, and as @dave-shirley-faraday points out, we already retry these for GET.

So let's merge the response->WaitStatus conversion and make it consistent.